### PR TITLE
[ui] Adds a "Scheduling" filter to the job.allocations page

### LINKED ISSUE #17221: Add a "Will not Reschedule" filter to the job/allocations page

### PR DESCRIPTION
**Pull request by philrenaud**
Adds a new filter, "Scheduling", to the job allocations page. This helps facilitate Rescheduled/Restarted links from the upcoming [Job Status Panel](https://github.com/hashicorp/nomad/issues/16128) and gives us future options for letting users drill down on this list page.

Because of some seen inconsistencies in testing, this adds `NextAllocation` to Allocations in their list/stub form from our API, where before they only existed when an allocation was being fetched individually. This, combined with a check for FollowUpEvalID, gives us a better picture of "Is a terminal allocation never going to be rescheduled?", which the UI is seeking to show to users here.

image

image

Resolves #17221

### CODE REVIEW

File: ui/app/components/job-status/panel/steady.hbs [hunk 1]
```diff
@@ -48,7 +48,7 @@
               @condition={{not (eq type.label "unplaced")}}
               @route="jobs.job.allocations"
               @model={{@job}}
-              @query={{hash status=(concat '["' type.label '"]') version=(concat '[' (keys this.versions) ']')}}
+              @query={{hash status=(concat '["' type.label '"]') version=(concat '[' (map-by "version" this.versions) ']')}}
```
Comment (philrenaud) [inline ui/app/components/job-status/panel/steady.hbs:51]: Noted that I was being inconsistent after changing my version sorting logic a few PRs ago.

**Review by ChaiWithJai — APPROVED** [review]

File: api/allocations.go [hunk 2]
```diff
@@ -326,6 +326,7 @@ func (a *Allocation) Stub() *AllocationListStub {
 		TaskStates:            a.TaskStates,
 		DeploymentStatus:      a.DeploymentStatus,
 		FollowupEvalID:        a.FollowupEvalID,
+		NextAllocation:        a.JobID,
```
Comment (lgfa29) [inline api/allocations.go:329]:
```suggestion
		NextAllocation:        a.NextAllocation,
```

File: ui/app/components/job-status/failed-or-lost.hbs [hunk 3]
```diff
@@ -9,14 +9,28 @@
...
+    >
+      {{@allocs.length}}
+    </ConditionalLinkTo>
+  {{/if}}
+  {{#if (eq @title "Restarted")}}
+    <ConditionalLinkTo
+      @condition={{this.shouldLinkToAllocations}}
+      @route="jobs.job.allocations"
+      @model={{@job}}
+      @query={{hash scheduling=(concat '["has-been-restarted"]') version=(concat '[' @job.latestDeployment.versionNumber ']')}}
```
Comment (lgfa29) [inline ui/app/components/job-status/failed-or-lost.hbs:29]: What is being concatened in `(concat '["has-been-rescheduled"]')` and `(concat '["has-been-restarted"]')`? It seems like it's just a string?

**Review by lgfa29 — CHANGES_REQUESTED** [review]
Just need to fix `api/allocations.go`, otherwise it will return the wrong value.

A `.changelog` entry would be good as well since it's a change to the API return content.

File: ui/app/components/job-status/failed-or-lost.hbs (hunk 3 above)
Comment (philrenaud) [reply ui/app/components/job-status/failed-or-lost.hbs:29]: Good catch; this was an artefact from earlier! Will fix.

**Pull request merged**